### DISCUSSION TITLE
git from browser

### DISCUSSION THREAD

**Discussion by nosovk** [sveltia/sveltia-cms#31] [Ideas]
Actually I found https://github.com/isomorphic-git/isomorphic-git
which could be useful for that project in some way.

**Comment by kyoshino** [comment]
That’s interesting! I’ll take a look at it. We could use it for complex Git operations that cannot be done by an API. While GitHub and GitLab offer an extensive REST/GraphQL API set, others may not.

* Gogs: I have never heard of it
* Gitea: Decap CMS is going to implement it as per https://github.com/decaporg/decap-cms/issues/1109
* Bitbucket: Can’t retrieve contents in bulk with the API
* Azure DevOps: Can’t retrieve contents in bulk with the API; requested in https://github.com/sveltia/sveltia-cms/discussions/2

**Comment by ngdangtu-vn** [comment]
Interesting idea, I wonder if we could convert git exe into webassembly (wasm) and use it instead of :?

**Comment by devurandom** [comment]
Support for Codeberg (uses Forgejo, a Gitea fork) would be interesting:
* https://codeberg.org/api/swagger
  - https://codeberg.org/api/swagger#/repository/repoCreateBranch
  - https://codeberg.org/api/swagger#/repository/repoChangeFiles
  - https://codeberg.org/api/swagger#/repository/repoCreatePullRequest
  - https://codeberg.org/api/swagger#/repository/repoMergePullRequest
* https://forgejo.org/docs/latest/user/api-usage/

**Reply by tobru** [reply]
I'd love to see support for Forgejo in Sveltia :+1:

**Reply by kyoshino** [reply]
Unfortunately we don’t have plan to support Gitea and Forgejo as mentioned in the README, as there is no way to fetch file contents at once.

**Comment by kyoshino** [comment]
I’ve just learned that `netlify-cms-proxy-server` has `git` mode that allows developers to create commits to a local repository. Maybe after the 1.0 release, I’ll figure out if [isomorphic-git](https://isomorphic-git.org/docs/en/0.78.0/fs) could be used with the local [File System Access API](https://developer.chrome.com/docs/capabilities/web-apis/file-system-access) to do the same thing.

- https://github.com/decaporg/decap-cms/issues/4429
- https://github.com/isomorphic-git/lightning-fs/issues/28
